generator/transit_generator.cpp

```diff
@@ -319,7 +208,179 @@ void BuildTransit(string const & mwmDir, string const & countryId,
+      ::IsValid(m_lines) && ::IsValid(m_shapes) && ::IsValid(m_networks);
+}
+
+void GraphData::SortStops() { sort(m_stops.begin(), m_stops.end(), LessById); }
+
+void GraphData::CalculateEdgeWeight()
+{
+  CHECK(is_sorted(m_stops.cbegin(), m_stops.cend(), LessById), ());
+
+  for (auto & e : m_edges)
+  {
+    if (e.GetWeight() != kInvalidWeight)
+      continue;
+
+    Stop const & s1 = FindStopById(m_stops, e.GetStop1Id());
+    Stop const & s2 = FindStopById(m_stops, e.GetStop2Id());
+    double const lengthInMeters = MercatorBounds::DistanceOnEarth(s1.GetPoint(), s2.GetPoint());
+    e.SetWeight(lengthInMeters / kTransitAverageSpeedMPS);
+  }
+}
+
+void GraphData::CalculateBestPedestrianSegment(string const & mwmPath, string const & countryId)
+{
+  // Creating IndexRouter.
+  SingleMwmIndex index(mwmPath);
+
+  auto infoGetter = storage::CountryInfoReader::CreateCountryInfoReader(GetPlatform());
+  CHECK(infoGetter, ());
+
+  auto const countryFileGetter = [&infoGetter](m2::PointD const & pt) {
+    return infoGetter->GetRegionCountryId(pt);
+  };
+
+  auto const getMwmRectByName = [&](string const & c) -> m2::RectD {
+    CHECK_EQUAL(countryId, c, ());
+    return infoGetter->GetLimitRectForLeaf(c);
+  };
+
+  CHECK_EQUAL(index.GetMwmId().GetInfo()->GetType(), MwmInfo::COUNTRY, ());
+  auto numMwmIds = make_shared<NumMwmIds>();
+  numMwmIds->RegisterFile(CountryFile(countryId));
+
+  // Note. |indexRouter| is valid while |index| is valid.
+  IndexRouter indexRouter(VehicleType::Pedestrian, false /* load altitudes */,
+                          CountryParentNameGetterFn(), countryFileGetter, getMwmRectByName,
+                          numMwmIds, MakeNumMwmTree(*numMwmIds, *infoGetter),
+                          traffic::TrafficCache(), index.GetIndex());
+
+  // Looking for the best segment for every gate.
+  for (auto & gate : m_gates)
+  {
+    // Note. For pedestrian routing all the segments are considered as twoway segments so
+    // IndexRouter.FindBestSegment() method finds the same segment for |isOutgoing| == true
+    // and |isOutgoing| == false.
+    Segment bestSegment;
+    try
+    {
+      // @todo(bykoianko) For every call of the method below WorldGraph is created. It's not
+      // efficient to create WorldGraph for every gate. The code should be redesigned.
+      if (indexRouter.FindBestSegmentInSingleMwm(gate.GetPoint(),
+                                                 m2::PointD::Zero() /* direction */,
+                                                 true /* isOutgoing */, bestSegment))
+      {
+        CHECK_EQUAL(bestSegment.GetMwmId(), 0, ());
+        gate.SetBestPedestrianSegment(SingleMwmSegment(
+            bestSegment.GetFeatureId(), bestSegment.GetSegmentIdx(), bestSegment.IsForward()));
+      }
+    }
+    catch (RootException const & e)
+    {
+      LOG(LDEBUG, ("Point of a gate belongs to several mwms or doesn't belong to any mwm. Gate:",
```

Comment:
LERROR please.

Reply:
Also, how about to return false or re-throw the exception?

Reply:
Pokemon exception handling? https://blog.codinghorror.com/new-programming-jargon/

How you may be sure that RootExceptions means only that "Point of a gate belongs to several mwms or ..."?

Reply:
The exception is raised by the following code:
```c++
bool IndexRouter::FindBestSegment(...l) const
{
  auto const file = platform::CountryFile(m_countryFileFn(point));
  MwmSet::MwmHandle handle = m_index.GetMwmHandleByCountryFile(file);
  if (!handle.IsAlive())
    MYTHROW(MwmIsNotAlifeException, ("Can't get mwm handle for", file));
...
```
So, |file| may be set to CountryFile("") if according to map data gates are located outside of simplified(!) mwm border. I write LOG(LERROR, ...) but are you sure that it's enough for LOG(LERROR, ...)?

Reply:
> Also, how about to return false or re-throw the exception?

It's a possible situation that the best segment is not found and it depends on the map. In that case I prefer keep building mwms.

Reply:
My main point is that error message may be wrong and misleading. Are you sure that this is the only place where exception is thrown now? If yes, how you can guarantee that in future there will be the only place where exception is thrown? And why do you catch RootException instead of RoutingException?

Reply:
> And why do you catch RootException instead of RoutingException? 

Please have a look at the code. I catch a special MwmIsNotAlifeException derived from RoutingException now. Besides that I rewrote the error message.

Reply:
In PR which is going after this one I check if points of gates lay within processed mwm before to call FindBestSegmentInSingleMwm(...) which may raise an exception. I wrote:
```c++
...
  for (auto & gate : m_gates)
  {
    if (countryFileGetter(gate.GetPoint()) != countryId)
      continue;
    try
    {
      if (indexRouter.FindBestSegmentInSingleMwm(..))
      {
        CHECK_EQUAL(bestSegment.GetMwmId(), 0, ());
        gate.SetBestPedestrianSegment(SingleMwmSegment(
            bestSegment.GetFeatureId(), bestSegment.GetSegmentIdx(), bestSegment.IsForward()));
      }
    }
    catch (MwmIsNotAliveException const & e)
    {
      LOG(LCRITICAL, ("Point of a gate belongs to the processed mwm:", countryId, ","
          "but the mwm is not alive. Gate:", gate, e.what()));
    }
...
```

Reply:
I did it in this PR